{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding files and data using Data Connect\n",
    "\n",
    "#### Learning Objectives\n",
    "Workshop attendees will learn how use the GA4GH Data Connect Service.  \n",
    "\n",
    "What will participants do as part of the exercise?\n",
    "\n",
    " - Understanding how to query data via Data Connect\n",
    " - Use Data Connect to find files that can be accessed via DRS\n",
    " - Learn how to obtain and use data descriptions (schema)\n",
    " - Discover the meaning of codes used in data\n",
    " \n",
    "\n",
    " #### Icons in this Guide\n",
    "\n",
    " üñê A hands-on section where you will code something or interact with the server\n",
    " \n",
    "### Query files\n",
    "The approach taken below is using mapping available through subject and specimen data available through the Data Connect API. \n",
    "\n",
    "Queries are submitted as SQL queries to one or more tables on the Data Connect server.\n",
    "\n",
    "As with other examples, first we set up a client to use the API. The server at DNAStack is used in the following examples.\n",
    "\n",
    "#### Step 1: Set up a Data Connect Client and run a predefined query "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "searchClient = DataConnectClient('https://data.publisher.dnastack.com/data-connect/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>bam_drs_id</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA18945</td>\n",
       "      <td>9327fb44eb81b49a41e38c8d86eb3b3a</td>\n",
       "      <td>SRR1601115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA18942</td>\n",
       "      <td>5c6244a1c2f499b356a1198b808383d9</td>\n",
       "      <td>SRR1604445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA18948</td>\n",
       "      <td>fb1cfb04d3ef99d07c21f9dbf87ccc68</td>\n",
       "      <td>SRR1601121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA18949</td>\n",
       "      <td>59dafcd573c26f7f481bcc2b72bda510</td>\n",
       "      <td>SRR1601123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA18947</td>\n",
       "      <td>4673e7afdcb89edf37fd974e1bf40e15</td>\n",
       "      <td>SRR1601119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NA19080</td>\n",
       "      <td>6f9f1fc52166530ed0568d61451b032f</td>\n",
       "      <td>SRR1598080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>NA19074</td>\n",
       "      <td>0805baa0849485a2a63ea41429b9b37c</td>\n",
       "      <td>SRR1604135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>NA19091</td>\n",
       "      <td>13ab0c2657909b758428821e85e5c91f</td>\n",
       "      <td>SRR1603948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>NA19086</td>\n",
       "      <td>1fe317586a8b8c2a8ca6dadebb9a0b4b</td>\n",
       "      <td>SRR1601181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>NA19085</td>\n",
       "      <td>f8a40ab899eabcc74c9410078c0d70af</td>\n",
       "      <td>SRR1598090</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows √ó 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name                        bam_drs_id         acc\n",
       "0       NA18945  9327fb44eb81b49a41e38c8d86eb3b3a  SRR1601115\n",
       "1       NA18942  5c6244a1c2f499b356a1198b808383d9  SRR1604445\n",
       "2       NA18948  fb1cfb04d3ef99d07c21f9dbf87ccc68  SRR1601121\n",
       "3       NA18949  59dafcd573c26f7f481bcc2b72bda510  SRR1601123\n",
       "4       NA18947  4673e7afdcb89edf37fd974e1bf40e15  SRR1601119\n",
       "..          ...                               ...         ...\n",
       "99      NA19080  6f9f1fc52166530ed0568d61451b032f  SRR1598080\n",
       "100     NA19074  0805baa0849485a2a63ea41429b9b37c  SRR1604135\n",
       "101     NA19091  13ab0c2657909b758428821e85e5c91f  SRR1603948\n",
       "102     NA19086  1fe317586a8b8c2a8ca6dadebb9a0b4b  SRR1601181\n",
       "103     NA19085  f8a40ab899eabcc74c9410078c0d70af  SRR1598090\n",
       "\n",
       "[104 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.search import DataConnectClient\n",
    "\n",
    "query = '''SELECT f.sample_name, drs_id bam_drs_id, acc\n",
    "FROM collections.public_datasets.onek_genomes_ssd_drs s \n",
    "join collections.public_datasets.onek_genomes_sra_drs_files f on f.sample_name = s.su_submitter_id \n",
    "where filetype = 'bam' and mapped = 'mapped' \n",
    "and sequencing_type ='exome' and  population = 'JPT' '''\n",
    "\n",
    "resultRows = searchClient.run_query(query, returnType='dataframe')\n",
    "resultRows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 2: Run a second query to find bam files from members of a given family"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>relationship</th>\n",
       "      <th>bam_drs_id</th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA12760</td>\n",
       "      <td>pat grandfather</td>\n",
       "      <td>a4ee62278644957c2f7ed39202184fdc</td>\n",
       "      <td>SRR1597880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA12762</td>\n",
       "      <td>mat grandfather</td>\n",
       "      <td>d3b77ffe47b452f92c061309ebab587d</td>\n",
       "      <td>SRR1601869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA12761</td>\n",
       "      <td>pat grandmother</td>\n",
       "      <td>57864dd794f647992c188dbea53a13df</td>\n",
       "      <td>SRR1601867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA12763</td>\n",
       "      <td>mat grandmother</td>\n",
       "      <td>988a24e166d8e103406cd62f7857b8b6</td>\n",
       "      <td>SRR1601870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name     relationship                        bam_drs_id         acc\n",
       "0     NA12760  pat grandfather  a4ee62278644957c2f7ed39202184fdc  SRR1597880\n",
       "1     NA12762  mat grandfather  d3b77ffe47b452f92c061309ebab587d  SRR1601869\n",
       "2     NA12761  pat grandmother  57864dd794f647992c188dbea53a13df  SRR1601867\n",
       "3     NA12763  mat grandmother  988a24e166d8e103406cd62f7857b8b6  SRR1601870"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "family_query = '''SELECT f.sample_name, relationship, drs_id bam_drs_id, acc\n",
    "FROM collections.public_datasets.onek_genomes_thousand_genomes_meta s \n",
    "join collections.public_datasets.onek_genomes_sra_drs_files f on f.sample_name = s.sample \n",
    "where filetype = 'bam' and mapped = 'mapped' \n",
    "and sequencing_type ='exome' and  family_id = '1447' '''\n",
    "\n",
    "family_results = searchClient.run_query(family_query, returnType='dataframe')\n",
    "family_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List table details\n",
    "\n",
    "\n",
    "#### Step 3:\n",
    "We can list the available tables available in this set as follows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the table list\n",
      "____Page1_______________\n",
      "thousand_genomes.onek_genomes.bdc_1000genomes\n",
      "thousand_genomes.onek_genomes.onek_drs\n",
      "thousand_genomes.onek_genomes.onek_recal_variants_drs\n",
      "thousand_genomes.onek_genomes.sra_drs_files\n",
      "thousand_genomes.onek_genomes.sra_onek_drs\n",
      "thousand_genomes.onek_genomes.ssd_drs\n",
      "thousand_genomes.onek_genomes.ssd_drs_copy\n",
      "thousand_genomes.onek_genomes.thousand_genomes_meta\n"
     ]
    }
   ],
   "source": [
    "table_list = searchClient.list_catalog('collections')  # This will list all accessible tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 4: List schema for sra_drs_files table\n",
    "The following cells can be run to list the columns for the other tables used in the queries above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablethousand_genomes.onek_genomes.sra_drs_files_\n",
      "{\n",
      "   \"name\": \"thousand_genomes.onek_genomes.sra_drs_files\",\n",
      "   \"description\": \"Thousand genomes Sequence Read Archive file data\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"TG_sra_t1\",\n",
      "      \"description\": \"Thousand genomes Sequence Read Archive file data\",\n",
      "      \"properties\": {\n",
      "         \"acc\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v1\",\n",
      "            \"description\": \"SRA run accession no\"\n",
      "         },\n",
      "         \"drs_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v3\",\n",
      "            \"description\": \"local DRS id unique within the DRS service\"\n",
      "         },\n",
      "         \"mapped\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v6\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"mapped\",\n",
      "                  \"title\": \"mapped to reference\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"unmapped\",\n",
      "                  \"title\": \"unmapped to reference\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Indicates whether the file contains reads which could be mapped to the reference genome, or reads which could not (unmapped). See https://www.internationalgenome.org/faq/what-are-unmapped-bams/\\t\"\n",
      "         },\n",
      "         \"filename\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v2\",\n",
      "            \"description\": \"Name of file in SRA\"\n",
      "         },\n",
      "         \"filetype\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v4\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"bam\",\n",
      "                  \"title\": \"Binary sequence alignment\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"bai\",\n",
      "                  \"title\": \"BAM index file\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Type of file\"\n",
      "         },\n",
      "         \"sample_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v5\",\n",
      "            \"description\": \"1000 genomes project sample name\"\n",
      "         },\n",
      "         \"sequencing_type\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v7\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"exome\",\n",
      "                  \"title\": \"Whole exome sequencing\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"low_coverage\",\n",
      "                  \"title\": \"Low coverage whole genome sequencing\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Type of genomic sequencing represented in the file\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "schema1 = searchClient.list_table_info('collections.public_datasets.onek_genomes_sra_drs_files', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_Schema for tablethousand_genomes.onek_genomes.ssd_drs_\n",
      "{\n",
      "   \"name\": \"thousand_genomes.onek_genomes.ssd_drs\",\n",
      "   \"description\": \"Thousand genomes sample metadata\",\n",
      "   \"data_model\": {\n",
      "      \"$id\": \"TG_sra_t2\",\n",
      "      \"description\": \"Thousand genomes sample metadata\",\n",
      "      \"properties\": {\n",
      "         \"md5sum\": {\n",
      "            \"type\": \"number\",\n",
      "            \"$id\": \"TG_sra_v21\",\n",
      "            \"description\": \"MD5 checksum for file\"\n",
      "         },\n",
      "         \"data_type\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v22\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"Aligned_20_Reads\",\n",
      "                  \"title\": \"Aligned_20_Reads\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Type of data contained in file\"\n",
      "         },\n",
      "         \"file_name\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v19\",\n",
      "            \"description\": \"Name of file\"\n",
      "         },\n",
      "         \"file_size\": {\n",
      "            \"type\": \"number\",\n",
      "            \"$id\": \"TG_sra_v20\",\n",
      "            \"$unit\": \"bytes\",\n",
      "            \"description\": \"Size of file in bytes\"\n",
      "         },\n",
      "         \"population\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v10\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"ACB\",\n",
      "                  \"title\": \"African Caribbean in Barbados\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ASW\",\n",
      "                  \"title\": \"African Ancestry in Southwest US\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"BEB\",\n",
      "                  \"title\": \"Bengali in Bangladesh\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CDX\",\n",
      "                  \"title\": \"Chinese Dai in Xishuangbanna, China\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CEU\",\n",
      "                  \"title\": \"Utah residents (CEPH) with Northern and Western European ancestry \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CHB\",\n",
      "                  \"title\": \"Han Chinese in Beijing, China\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CHS\",\n",
      "                  \"title\": \"Han Chinese South\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"CLM\",\n",
      "                  \"title\": \"Colombian in Medellin, Colombia\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ESN\",\n",
      "                  \"title\": \"Esan in Nigeria\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"FIN\",\n",
      "                  \"title\": \"Finnish in Finland \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GBR\",\n",
      "                  \"title\": \"British in England and Scotland \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GIH\",\n",
      "                  \"title\": \"Gujarati Indian in Houston, TX\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"GWD\",\n",
      "                  \"title\": \"Gambian in Western Division, The Gambia \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"IBS\",\n",
      "                  \"title\": \"Iberian populations in Spain \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"ITU\",\n",
      "                  \"title\": \"Indian Telugu in the UK\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"JPT\",\n",
      "                  \"title\": \"Japanese in Tokyo, Japan\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"KHV\",\n",
      "                  \"title\": \"Kinh in Ho Chi Minh City, Vietnam\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"LWK\",\n",
      "                  \"title\": \"Luhya in Webuye, Kenya\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"MSL\",\n",
      "                  \"title\": \"Mende in Sierra Leone\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"MXL\",\n",
      "                  \"title\": \"Mexican Ancestry in Los Angeles, California\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PEL\",\n",
      "                  \"title\": \"Peruvian in Lima, Peru\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PJL\",\n",
      "                  \"title\": \"Punjabi in Lahore, Pakistan\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"PUR\",\n",
      "                  \"title\": \"Puerto Rican in Puerto Rico\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"STU\",\n",
      "                  \"title\": \"Sri Lankan Tamil in the UK\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"TSI\",\n",
      "                  \"title\": \"Toscani in Italia \"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"YRI\",\n",
      "                  \"title\": \"Yoruba in Ibadan, Nigeria\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Population list\"\n",
      "         },\n",
      "         \"data_format\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v23\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"CRAM\",\n",
      "                  \"title\": \"CRAM\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Format of data file\"\n",
      "         },\n",
      "         \"read_drs_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v18\",\n",
      "            \"description\": \"DRS id for read file\"\n",
      "         },\n",
      "         \"annotated_sex\": {\n",
      "            \"type\": \"encoded value\",\n",
      "            \"$id\": \"TG_sra_v11\",\n",
      "            \"oneOf\": [\n",
      "               {\n",
      "                  \"const\": \"male\",\n",
      "                  \"title\": \"male\"\n",
      "               },\n",
      "               {\n",
      "                  \"const\": \"female\",\n",
      "                  \"title\": \"female\"\n",
      "               }\n",
      "            ],\n",
      "            \"description\": \"Sex of participant\"\n",
      "         },\n",
      "         \"al_submitter_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v16\",\n",
      "            \"description\": \"Aliquot id used by submitter\"\n",
      "         },\n",
      "         \"rg_submitter_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v17\",\n",
      "            \"description\": \"Read group id used by submitter\"\n",
      "         },\n",
      "         \"sa_submitter_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v15\",\n",
      "            \"description\": \"Sample id used by submitter\"\n",
      "         },\n",
      "         \"su_submitter_id\": {\n",
      "            \"type\": \"string\",\n",
      "            \"$id\": \"TG_sra_v14\",\n",
      "            \"description\": \"Subject id used by submitter\"\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "schema2 = searchClient.list_table_info('collections.public_datasets.onek_genomes_ssd_drs', verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 6: Search for a different population group \n",
    "üñê Using the information above about the tables, modify the query to use\n",
    "a) a population code that represents Gujarati Indians living in a city in Texas.\n",
    "b) bam files for reads that have not been mapped to a reference genome.\n",
    "\n",
    "So you don't have to modify the sql query itself you can add the values you identified to the variables in the next cell of this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "population_code = 'GIH'\n",
    "mapping_type = 'unmapped'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving the query\n",
      "____Page1_______________\n",
      "____Page2_______________\n",
      "____Page3_______________\n",
      "____Page4_______________\n",
      "____Page5_______________\n",
      "____Page6_______________\n",
      "____Page7_______________\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>bam_drs_id</th>\n",
       "      <th>acc</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA20853</td>\n",
       "      <td>1f0a7f9e10ebf38d45bbb08cffe69851</td>\n",
       "      <td>SRR1598480</td>\n",
       "      <td>NA20853.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA20852</td>\n",
       "      <td>b55c589bfcabc81355b2fdbb2f81ff37</td>\n",
       "      <td>SRR1598478</td>\n",
       "      <td>NA20852.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA20894</td>\n",
       "      <td>dcdd7fcafa21f5d69a0ab650114faf88</td>\n",
       "      <td>SRR1598538</td>\n",
       "      <td>NA20894.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA20895</td>\n",
       "      <td>7c45236b410cef0bc85e6052f810ef69</td>\n",
       "      <td>SRR1598540</td>\n",
       "      <td>NA20895.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA20899</td>\n",
       "      <td>2a8edfc0ffaa964348fbbc0c15d7231c</td>\n",
       "      <td>SRR1598548</td>\n",
       "      <td>NA20899.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>NA21114</td>\n",
       "      <td>bc21dbc917062f4c6e79e08f00bc2bcf</td>\n",
       "      <td>SRR1598598</td>\n",
       "      <td>NA21114.unmapped.ILLUMINA.bwa.GIH.exome.201205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NA21107</td>\n",
       "      <td>8bd32cc867deeab6c8d5e9fd4eca8ff6</td>\n",
       "      <td>SRR1600518</td>\n",
       "      <td>NA21107.unmapped.ILLUMINA.bwa.GIH.exome.201304...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>NA21111</td>\n",
       "      <td>928ccbddac73668de5b61d4bb4d0df4e</td>\n",
       "      <td>SRR1600526</td>\n",
       "      <td>NA21111.unmapped.ILLUMINA.bwa.GIH.exome.201304...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>NA21108</td>\n",
       "      <td>fd0bdfa174cf6e721c4b3921bb840cfc</td>\n",
       "      <td>SRR1602948</td>\n",
       "      <td>NA21108.unmapped.ILLUMINA.bwa.GIH.exome.201304...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>NA21109</td>\n",
       "      <td>c5386a8f71a94142e0fe02e12f0a086c</td>\n",
       "      <td>SRR1604340</td>\n",
       "      <td>NA21109.unmapped.ILLUMINA.bwa.GIH.exome.201304...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name                        bam_drs_id         acc  \\\n",
       "0       NA20853  1f0a7f9e10ebf38d45bbb08cffe69851  SRR1598480   \n",
       "1       NA20852  b55c589bfcabc81355b2fdbb2f81ff37  SRR1598478   \n",
       "2       NA20894  dcdd7fcafa21f5d69a0ab650114faf88  SRR1598538   \n",
       "3       NA20895  7c45236b410cef0bc85e6052f810ef69  SRR1598540   \n",
       "4       NA20899  2a8edfc0ffaa964348fbbc0c15d7231c  SRR1598548   \n",
       "..          ...                               ...         ...   \n",
       "98      NA21114  bc21dbc917062f4c6e79e08f00bc2bcf  SRR1598598   \n",
       "99      NA21107  8bd32cc867deeab6c8d5e9fd4eca8ff6  SRR1600518   \n",
       "100     NA21111  928ccbddac73668de5b61d4bb4d0df4e  SRR1600526   \n",
       "101     NA21108  fd0bdfa174cf6e721c4b3921bb840cfc  SRR1602948   \n",
       "102     NA21109  c5386a8f71a94142e0fe02e12f0a086c  SRR1604340   \n",
       "\n",
       "                                              filename  \n",
       "0    NA20853.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "1    NA20852.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "2    NA20894.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "3    NA20895.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "4    NA20899.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "..                                                 ...  \n",
       "98   NA21114.unmapped.ILLUMINA.bwa.GIH.exome.201205...  \n",
       "99   NA21107.unmapped.ILLUMINA.bwa.GIH.exome.201304...  \n",
       "100  NA21111.unmapped.ILLUMINA.bwa.GIH.exome.201304...  \n",
       "101  NA21108.unmapped.ILLUMINA.bwa.GIH.exome.201304...  \n",
       "102  NA21109.unmapped.ILLUMINA.bwa.GIH.exome.201304...  \n",
       "\n",
       "[103 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = f'''SELECT f.sample_name, drs_id bam_drs_id, acc, filename\n",
    "FROM collections.public_datasets.onek_genomes_ssd_drs s \n",
    "join collections.public_datasets.onek_genomes_sra_drs_files f on f.sample_name = s.su_submitter_id \n",
    "\n",
    "where filetype = 'bam' and mapped = '{mapping_type}' \n",
    "and sequencing_type ='exome' and  population = '{population_code}'\n",
    "'''\n",
    "\n",
    "resultRows = searchClient.run_query(query, returnType='dataframe')\n",
    "resultRows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Important note\n",
    "Looking up a data dictionary to discover codes in this way is not what we would typically expect a user to do. Our aim today is to focus on the API and what it is capable of and what it can enable.\n",
    "\n",
    "Given the information the data schema provide about the data it is possible for developers to create interfaces in their systems which allow new datasources to be integrated as they appear.\n",
    "\n",
    "In another notebook (next). We'll look at an example of how a more user friendly user interface can be provided using the information that the the API provides."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 7 - Combine with DRS Server\n",
    "\n",
    "The following shows how the SRA DRS server we used in workbook 2-1 can be used to determine where the files we discovered can be obtained from. \n",
    "\n",
    "üñê Using the results from one of the queries that you ran above take a DRS id from the query results and use it in the following calls to the NCBI DRS server."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://locate.be-md.ncbi.nlm.nih.gov/ga4gh/drs/v1/objects/1f0a7f9e10ebf38d45bbb08cffe69851\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'access_methods': [{'access_id': '6d8f0e53659a16fba26e9abc4b40d3930929c448bd54f76398dac8788c90bf56',\n",
       "   'region': 'gs.US',\n",
       "   'type': 'https'},\n",
       "  {'access_id': '01894b008317005f919216578ffd8ef095c30d4541489f3db35e358bb74c7450',\n",
       "   'type': 'https'},\n",
       "  {'access_id': '5910ee14f991a7db0fa75bd24a4c5e169c206c09d9787e0fbed6a5ae4765b9a9',\n",
       "   'region': 's3.us-east-1',\n",
       "   'type': 'https'}],\n",
       " 'checksums': [{'checksum': '1f0a7f9e10ebf38d45bbb08cffe69851',\n",
       "   'type': 'md5'}],\n",
       " 'created_time': '2012-11-18T13:03:48Z',\n",
       " 'id': '1f0a7f9e10ebf38d45bbb08cffe69851',\n",
       " 'name': 'NA20853.unmapped.ILLUMINA.bwa.GIH.exome.20120522.bam',\n",
       " 'self_url': 'drs://locate.be-md.ncbi.nlm.nih.gov/1f0a7f9e10ebf38d45bbb08cffe69851',\n",
       " 'size': 114284941}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fasp.loc import DRSClient\n",
    "\n",
    "drsClient = DRSClient('https://locate.be-md.ncbi.nlm.nih.gov', public=True, debug=True)\n",
    "test_id = '1f0a7f9e10ebf38d45bbb08cffe69851'\n",
    "objInfo = drsClient.get_object(test_id)\n",
    "objInfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A second DRS call can be used to obtain a url to access the file from one of the above locations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "access_id:6d8f0e53659a16fba26e9abc4b40d3930929c448bd54f76398dac8788c90bf56\n",
      "https://locate.be-md.ncbi.nlm.nih.gov/ga4gh/drs/v1/objects/1f0a7f9e10ebf38d45bbb08cffe69851/access/6d8f0e53659a16fba26e9abc4b40d3930929c448bd54f76398dac8788c90bf56\n",
      "<Response [200]>\n",
      "url:https://storage.googleapis.com/genomics-public-data/ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase3/data/NA20853/exome_alignment/NA20853.unmapped.ILLUMINA.bwa.GIH.exome.20120522.bam\n"
     ]
    }
   ],
   "source": [
    "access_id = objInfo['access_methods'][0]['access_id']\n",
    "print('access_id:{}'.format(access_id))\n",
    "url = drsClient.get_access_url(test_id, access_id=access_id)\n",
    "print('url:{}'.format(url))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
